File: ndb-bindings-0.7.1/java/test/java/testsuite/ndbj/OutOfConnectionProblemTest.java
/* -*- mode: java; c-basic-offset: 4; indent-tabs-mode: nil; -*-
 *  vim:expandtab:shiftwidth=4:tabstop=4:smarttab:
 *
 *  ndb-bindings: Bindings for the NDB API
 *  Copyright (C) 2008 MySQL
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package testsuite.ndbj;

import testsuite.BaseNdbjTestCase;
import com.mysql.cluster.ndbj.*;

public class OutOfConnectionProblemTest extends BaseNdbjTestCase {

    static boolean alreadySetUp = false;

    /////////////////////////////////////////////////////////////////////
    // GENERAL STUFF
    /////////////////////////////////////////////////////////////////////

    static final int     NUMBER_OF_TEST_ITERATIONS    = 1000000;

    static final int     NUMBER_OF_ELEMENT_INSERTIONS = 25;

    static final int     NUM_TRANSACTIONS             = 500;
    static final int     TRANSACTIONS_TIMEOUT         = 10;
    static final int     FORCE_SEND                   = 1;

    /////////////////////////////////////////////////////////////////////
    // TABLE METADATA
    /////////////////////////////////////////////////////////////////////

    static final String PROBTABLE  = "OutOfConnectionProblemTable";

    static int PROBTABLE_FIELD1;

    /////////////////////////////////////////////////////////////////////
    // OTHER
    /////////////////////////////////////////////////////////////////////

    static int myTransactionCount      = 0;
    static int myTotalTransactionCount = 0;

    /////////////////////////////////////////////////////////////////////
    // INITIALISATION CODE
    /////////////////////////////////////////////////////////////////////

    public static void main(String[] args) {
        junit.textui.TestRunner.run(OutOfConnectionProblemTest.class);
    }

    public OutOfConnectionProblemTest(String arg0) {
        super(arg0);
    }

    @Override
    public void setUp() throws Exception {
        super.setUp();

        if (!alreadySetUp) {
            // Create required tables

            createTable(PROBTABLE, "(field1 INT UNSIGNED NOT NULL, a integer, b integer, c integer, d integer, e int,f integer, g integer, PRIMARY KEY(field1)) ENGINE=ndb PARTITION BY KEY (field1);");

            NdbDictionary dictionary = ndb.getDictionary();

            NdbTable  probTable = dictionary.getTable(PROBTABLE);
            PROBTABLE_FIELD1    = probTable.getColumn("field1").getColumnNo();
        }
    }

    /////////////////////////////////////////////////////////////////////
    // TEST CODE
    /////////////////////////////////////////////////////////////////////

    public void testOperation()  throws NdbApiException {

        try {
            long startTime = System.currentTimeMillis();

            for (int iteration = 0;
                 iteration < NUMBER_OF_TEST_ITERATIONS;
                 iteration++) {
                generateRequestTxs(iteration);

                sendPollNdb(TRANSACTIONS_TIMEOUT, NUM_TRANSACTIONS, FORCE_SEND);
                System.gc(); System.runFinalization();
            }

            System.out.println("Performing cooldown");

            pollNdb(100000, myTransactionCount);

            long endTime     = System.currentTimeMillis();
            long elapsedTime = endTime - startTime;

            System.out.println("Test completed with no exceptions and took " + elapsedTime / 1000 + " seconds");
        }
        finally {
            System.out.println("End of test.");
        }
    }
    static int cnt=0;
    public void generateRequestTxs(int theLastBatch) throws NdbApiException {

        boolean  success = false;

        NdbOperation orderOp;
        BaseCallback callback;

        try {

            for (int elementCount = 1;
                 elementCount < (NUMBER_OF_ELEMENT_INSERTIONS) + 1;
                 elementCount++) {

                cnt++;
                trans = startCountedTransaction();

                orderOp = trans.getInsertOperation(PROBTABLE);

                orderOp.equalInt(PROBTABLE_FIELD1, cnt);
                orderOp.setInt("a", cnt);
                orderOp.setInt("b", cnt);
                orderOp.setInt("c", cnt);
                orderOp.setInt("d", cnt);
                orderOp.setInt("e", cnt);
                orderOp.setInt("f", cnt);
                orderOp.setInt("g", cnt);

                callback = new CountingCallback(cnt, ndb, orderOp.resultData(), trans);

                trans.executeAsynchPrepare(NdbTransaction.ExecType.Commit,
                                           callback,
                                           NdbOperation.AbortOption.AbortOnError);
            }


            success = true;
        }
        finally {
            System.out.println("Started " + myTotalTransactionCount + " txs in total " + (success ? "(no errors)" : "(errors, aborting)"));
        }
    }

    private NdbTransaction startCountedTransaction() throws NdbApiException {
        myTransactionCount++;
        myTotalTransactionCount++;

        return ndb.startTransaction();
    }

    private int sendPollNdb(int aMillisecondNumber, int minNoOfEventsToWakeup, int forceSend) throws NdbApiException {
        ndb.sendPreparedTransactions(forceSend);

        return pollNdb(aMillisecondNumber, minNoOfEventsToWakeup);
    }

    private int pollNdb(int aMillisecondNumber, int minNoOfEventsToWakeup) throws NdbApiException {
        int numberOfCallbacks = ndb.pollNdb(aMillisecondNumber, minNoOfEventsToWakeup);

        System.out.println("Poll produced " + numberOfCallbacks + " callbacks (" + myTransactionCount +
                           " txs outstanding) [Total " + myTotalTransactionCount + "]");

        return numberOfCallbacks;
    }

    class CountingCallback extends BaseCallback {
        NdbTransaction t;
        public CountingCallback(int theCallbackNum, Ndb theNdb, NdbResultSet theResults, NdbTransaction trans) {
            super(trans);
            t=trans;
        }

        @Override
        public void callback(int theResult) {
            myTransactionCount--;
            //System.out.print("Callback (" + theTrans + "): ");

            //System.out.print("Transaction was " + (theTrans.isClosed() ? "closed" : "open") + " (before close()) " );
            //theTrans.close();
            t.close();

            //System.out.println("and is now " + (theTrans.isClosed() ? "closed" : "open") );
        }

    }

}

File: ndb-bindings-0.7.1/cpp/OutOfConnectionProblemTest.cpp




#include <mysql.h>
#include <NdbApi.hpp>

// Used for cout
#include <stdio.h>
#include <iostream>
#include <time.h>


#define NUMBER_OF_TEST_ITERATIONS  1000000
#define NUMBER_OF_ELEMENT_INSERTIONS 25

#define NUM_TRANSACTIONS             500
#define TRANSACTIONS_TIMEOUT         10
#define FORCE_SEND                   1

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////
	// TABLE METADATA
	///////////////////////////////////////////////////////////////////////////////////////////////////////////////

#define PROBTABLE  "OutOfConnectionProblemTable"

int PROBTABLE_FIELD1;

int myTransactionCount      = 0;
int myTotalTransactionCount = 0;


#define PRINT_ERROR(code,msg) \
  std::cout << "Error in " << __FILE__ << ", line: " << __LINE__ \
            << ", code: " << code \
            << ", msg: " << msg << "." << std::endl
#define MYSQLERROR(mysql) { \
  PRINT_ERROR(mysql_errno(&mysql),mysql_error(&mysql)); \
  exit(-1); }
#define APIERROR(error) { \
  PRINT_ERROR(error.code,error.message); \
  exit(-1); }

int random_id = 0;

#define NUM_ITER 1000
#define INSERT_NUM 1000

typedef struct
async_callback_t
{
  Ndb *ndb;
  int cnt; 
};
static void callBack(int result,
                     NdbTransaction *trans,
                     void *aObject)
{
  async_callback_t *cbData((async_callback_t *)aObject);
  int tid= cbData->cnt;
  myTransactionCount--;
  if (result < 0)
  {
    if(trans!=0)
    {
      std::cout << "error!! Error code: " << trans->getNdbError().code
           << "Err message: " << trans->getNdbError().message
           << std::endl;
      trans->close();
      return;
    }
  }
  if(trans!=0)
  {
    trans->close();
  }
  free(cbData); 
}

int cnt=0;
Ndb * ndb = NULL; 

NdbTransaction * startCountedTransaction() {
  myTransactionCount++;
  myTotalTransactionCount++;
  
  return ndb->startTransaction();
}

int pollNdb(int aMillisecondNumber, int minNoOfEventsToWakeup) {
  int numberOfCallbacks = ndb->pollNdb(aMillisecondNumber, minNoOfEventsToWakeup);
  
  std::cout << "Poll produced " << numberOfCallbacks << " callbacks (" << myTransactionCount; 
  std::cout << " txs outstanding) [Total " <<  myTotalTransactionCount << "]" <<std::endl;
  
  return numberOfCallbacks;
}

int sendPollNdb(int aMillisecondNumber, int minNoOfEventsToWakeup, int forceSend) {
  ndb->sendPreparedTransactions(forceSend);
  
  return pollNdb(aMillisecondNumber, minNoOfEventsToWakeup);
}

void generateRequestTxs(int theLastBatch) {
  
  bool  success = false;
  
  NdbOperation * orderOp;
  async_callback_t * callback_data; 
  
  
  for (int elementCount = 1;
       elementCount < (NUMBER_OF_ELEMENT_INSERTIONS) + 1;
       elementCount++) {
    
    cnt++;
    NdbTransaction * trans = startCountedTransaction();
    
    orderOp = trans->getNdbOperation(PROBTABLE);
    
    orderOp->insertTuple();
    
    orderOp->equal(PROBTABLE_FIELD1, cnt);
    orderOp->setValue("a", cnt);
    orderOp->setValue("b", cnt);
    orderOp->setValue("c", cnt);
    orderOp->setValue("d", cnt);
    orderOp->setValue("e", cnt);
    orderOp->setValue("f", cnt);
    orderOp->setValue("g", cnt);
    
    callback_data = (async_callback_t *)malloc(sizeof(async_callback_t));
    
    
    callback_data->ndb=ndb; 
    callback_data->cnt=cnt;
    
    
    trans->executeAsynchPrepare(NdbTransaction::Commit,&callBack,
				(void *)callback_data, 
				NdbOperation::AbortOnError);
    


    success = true;
    
  }
  
}


int main()
{
  ndb_init();
  MYSQL mysql;

  /**************************************************************
   * Connect to mysql server and create table                   *
   **************************************************************/

  {
    if ( !mysql_init(&mysql) ) {
      std::cout << "mysql_init failed\n";
      exit(-1);
    }
    mysql_options(&mysql,MYSQL_READ_DEFAULT_GROUP,"client");
    if ( !mysql_real_connect(&mysql, "localhost", "root", "", "",
			     3306, NULL, 0) )
      MYSQLERROR(mysql);

    if (mysql_query(&mysql, "USE test") != 0) MYSQLERROR(mysql);

    if (mysql_query(&mysql, 
		    "CREATE TABLE if not exists "
		    PROBTABLE
		    " (field1 INT UNSIGNED NOT NULL, "
		    "a integer, "
		    "b integer, "
		    "c integer, "
		    "d integer, "
		    "e int, "
		    "f integer, "
		    "g integer, "
		    "PRIMARY KEY(field1)) ENGINE=ndb "
		    "PARTITION BY KEY (field1);"))
      MYSQLERROR(mysql);
  }

  /**************************************************************
   * Connect to ndb cluster                                     *
   **************************************************************/

  Ndb_cluster_connection *cluster_connection=
    new Ndb_cluster_connection(); // Object representing the cluster

  if (cluster_connection->connect(5,3,1))
  {
    std::cout << "Connect to cluster management server failed.\n";
    exit(-1);
  }

  if (cluster_connection->wait_until_ready(30,30))
  {
    std::cout << "Cluster was not ready within 30 secs.\n";
    exit(-1);
  }


  ndb = new Ndb( cluster_connection,
			"test" );  // Object representing the database
  if (ndb->init(1000) == -1) { 
    APIERROR(ndb->getNdbError());
    exit(-1);
  }

  const NdbDictionary::Dictionary* myDict= ndb->getDictionary();
  const NdbDictionary::Table *myTable= myDict->getTable(PROBTABLE);
  PROBTABLE_FIELD1    = myTable->getColumn("field1")->getColumnNo();
  
  if (myTable == NULL)
    APIERROR(myDict->getNdbError());


  
  for (int iteration = 0;
       iteration < NUMBER_OF_TEST_ITERATIONS;
       iteration++) {
    generateRequestTxs(iteration);
    
    sendPollNdb(TRANSACTIONS_TIMEOUT, NUM_TRANSACTIONS, FORCE_SEND);	
  }

  std::cout << "Performing cooldown" << std::endl;

  pollNdb(100000, myTransactionCount);	


  std::cout << "Test completed with no exceptions" << std::endl;

  

}
